---
title: ""
author: "Josefin"
output: 
  html_document:
    toc: true
    toc_float: true
    css: "style.css"
    includes:
      in_header: header.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(results='hide', eval=FALSE, echo=TRUE)
```

## Supplementary tables

* Supplementary Table S1: Master metadata file. Includes metadata derived from BTyperDB, IGUA clustering results, biosynthetic classes and much more.
* Supplementary Table S2: Contains GCF0001397-associated metadata, extracted from Supplementary Table S1 and NCBI.

Please note that the figure code below assumes that the headers of the supplementary tables have been removed prior to running the R code.

## Seurat object

* `btyperbgc_v1_0.55res_w_metadata.RDS` - Seurat object associated with BTyperBGC (v1). This UMAP is the underlying basis of the associated Shiny application (see: https://github.com/c20josbl/btyperbgc_shiny).

## Additional scripts and code

* `bacillus_parser.py` - Python script used to append "AS." to the start of antiSMASH BGCs, as well as ensure that a full, identifiable name was present in the LOCUS field of the GenBank files.
* `umap.py`- was used to annotate Pfam domains among GCF representatives.
* `parsing_classes.py` - Python script used to parse biosynthetic classes from the GenBank file. Note that this script fails to extract biosynthetic classes from the MIBiG (v4.0) BGCs, due to them not being present in the GenBank files.

Please note that Nextflow was used to manage job submissions when running antiSMASH and GECCO on a cluster. The code samples below are examples with identifiable paths removed.

```{r gecco_code}
gecco run -j 2 --genome genome.fa --o output_dir --mask --threshold 0.3 --merge-gbk
```

```{r}
antismash -c 2 --output-dir output_dir --output-basename genome_prefix --genefinding-tool prodigal-m genome.fa
```

antiSMASH, GECCO and MIBiG (v4.0) BGCs were concatenated to one GenBank file, which was then used as input for clustering by IGUA.

```{r}
igua -i bacillus_total_mibig.gbk -o btyperbgc_v2_igua.tsv -j 28 --workdir temp --clustering-distance 0.80 --clustering-method "average"
```

The LoVis4u figures were generated by running commands similar to the example below:
```{r}
lovis4u -gb folder_with_genbanks -hl -scc --reorient_loci --run-hmmscan --show-all-feature-labels -o output_dir_name --set-group-colour-for conserved --figure-width 110
```

For the full GCF figure, the options `--set-group-colour-for conserved` and `--figure-width 110` were not included, and instead the option `--use-filename-as-id` was added. 

For the maxmimum-likelihood tree, the following command was used to create `.gff` files from the `.fna` files via Prokka (v1.14.6):

```{r}
prokka --addgenes --outdir outdir_name
```

The core alignment was produced using Panaroo (v1.3.4), with `files.log` containing paths to the created `.gff` files:

```{r}
panaroo -i files.log -o outdir_name --clean-mode strict --core_threshold 0.95 --remove-invalid-genes -f 0.5 --aligner mafft
panaroo-msa -t 28 -o outdir_name -a core
```

Constant sites were filtered from the core gene alignment using snp-sites (v2.5.1):

```{r}
snp-sites -c -o snp_output panaroo_outdir_name/core_gene_alignment_filtered.aln
snp-sites -C panaroo_outdir_name/core_gene_alignment_filtered.aln
```

The tree files were subsequently produced via IQ-TREE (v2.2.5). A *B. bingmayongensis* genome was set as the outgroup. Replace the XXX with the output from running the second snp-sites command above:

```{r}
iqtree -s snp_output -m GTR+I+R -fconst XXX,XXX,XXX,XXX --seed 1001 --prefix outputfiles_iqtree -nt 28 -B 1000 -o BTDB_2022-0000619.2
```

A custom BLASTp database was created using MIBiG (v3.1) BGC files and command line BLAST (v2.14.1). Subsequently, a GCF0001397 representative was used to query this database. Sample commands for this are provided below.

```{r}
makeblastdb -in mibig_cds.fasta -dbtype prot

blastp -query GCF0001397_representative.fasta -db mibig_cds.fasta -out output.tsv -outfmt "6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp"
```


## Figure code

Code for all figures included in the manuscript has been included below. Please note that the code chunks will not executed when knitting the notebook, and thus the generated HTML file will only include the code chunks. For versions of R and packages that were used when running the original analysis, please see the full manuscript.

```{r, results='hide', eval=FALSE, echo=TRUE}

#Packages used throughout this notebook

library(tidyverse)
library(Seurat)
library(Signac)
library(clustree)
library(ggridges)
library(ggpubr)
library(ape)
library(phytools)
library(ggtree)
library(ggplotify)
library(VennDiagram)


```


```{r}
##FIGURE 2 MASTER CODE

#Panel A

#Importing the Seurat object and relevant metadata

pfams <- readRDS("btyperbgc_v1_0.55res_w_metadata_only_high.RDS")

metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Assigning a Novelty column

patterns=c("BTDB", "AS.")

metadata=metadata %>% 
  group_by(gcf_id) %>% 
  mutate(
    Novelty=ifelse(all(grepl('^BGC', cluster_id)), "MIBiG",
                ifelse(all(grepl(paste(patterns, collapse='|'), cluster_id)),"Novel",
                              "Mix")
                )) %>%
  ungroup() %>% 
  select(cluster_id, Novelty)

metadata=as.data.frame(metadata)

rownames(metadata)=metadata$cluster_id

#Attaching metadata with Novelty information to the SeuratObject

pfams=AddMetaData(pfams, metadata = metadata)

tols_cols=c("#66CCEE", "#EE6677","#4477AA")

#Making the plot

p_umap_novelty=DimPlot(pfams, group.by = "Novelty", order=TRUE, raster=FALSE,
        pt.size = 2.5, alpha=0.5, cols=tols_cols)+
  ggtitle("")


#Panel B

#Setting up dataframe

df_novelty_barplot=data.frame(
  name=c("Novel", "Mix", "MIBiG"),
  value=c(78.23125, 20.49972, 1.269031)
)

#Making the bar plot

p_bar_novelty=ggplot(df_novelty_barplot, aes(x=as.factor(reorder(name,-value)), y=value,
               fill=name, colour=name))+
  geom_bar(stat = "identity")+
  theme_classic()+
  xlab("")+
  ylab("% of full BGC dataset")+
  ylim(0,100)+
  scale_fill_manual(values = c("#66CCEE", "#EE6677", "#4477AA"))+
  scale_colour_manual(values = c("black", "black", "black"))+
  theme(legend.position = "none")

#Panel C

#Importing supplementary table S1

metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Retaining GCF containing at least 1 MIBiG BGC
mibig_gcfs=metadata %>% 
  filter(grepl('MIBiG', Tool))

bgc_types_mibig_gcfs=mibig_gcfs %>% 
  mutate(
    Type=ifelse(grepl('^BGC', cluster_id), "MIBiG",
                ifelse(grepl('^BTDB', cluster_id),"GECCO",
                       "antiSMASH")
                ))

bgc_types_mibig_gcfs=bgc_types_mibig_gcfs %>% 
  group_by(gcf_id) %>% 
  filter(n_distinct(Type) > 1) %>% 
  ungroup()

#Setting up dataframe for stacked bar plot

stacked_barplot_df=bgc_types_mibig_gcfs %>% 
  group_by(gcf_id, Type) %>% 
  summarise(n=n())

#Making the stacked bar plot

p_stacked_bar_novelty=ggplot(stacked_barplot_df, aes(fill=Type, y=n, x=reorder(gcf_id,-n)))+
  geom_bar(position = "stack", stat = "identity")+
  theme_classic()+
  theme(axis.text.x = element_text(angle=90))+
  ylab("# of BGCs")+
  xlab("")+
  scale_color_manual(values = c("#AA3377", "#228833", "#66CCEE"))+
  scale_fill_manual(values = c("#AA3377", "#228833", "#66CCEE"))

#Making the final panel figure

ggarrange(p_umap_novelty,
          ggarrange(p_bar_novelty, p_stacked_bar_novelty, align="hv", nrow = 2, heights = c(1.6,2), labels = c("B", "C")),
          align="hv", widths = c(1.3,1), labels = c("A", "B"))


ggsave('fig_2.png', dpi = 350, height = 7.5, width = 15)
```

```{r}
## FIGURE 3 MASTER CODE

#Importing supplementary table S1
metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Setting up a dataframe for the ridge plot

bacillus=data.frame(metadata$BTyperDB_ID,metadata$gcf_id,
                    metadata$BTyper3_Adjusted_panC_Group.predicted_species.)
colnames(bacillus)=c("BTyperDB_ID","GCF_ID","panC")

filtered=bacillus %>% 
  group_by(panC,
           BTyperDB_ID) %>% 
  summarise(n=n())

filtered$panC <- ifelse(grepl("\\*$", filtered$panC), "Unknown", filtered$panC)

filtered$panC=sapply(filtered$panC, function(x)
  sub("Unknown(missing allele)", "Unknown",x,fixed=TRUE))

filtered=filtered %>% 
  filter(!grepl('^BGC', BTyperDB_ID))

filtered_2=filtered %>% 
  filter(!grepl('^Group_bingmayongensis', panC)) %>% 
  filter(!grepl('^Group_clarus', panC)) %>% 
  filter(!grepl('^Group_gaemo', panC)) %>% 
  filter(!grepl('^Group_manliponensis', panC))

#Making the plot

ggplot(filtered_2, aes(x=n, y=panC, fill=panC)) +
  geom_density_ridges() +
  theme_ridges() + 
  theme(legend.position = "none",
        axis.title.y = element_blank())+
  xlab("# of GCFs per genome")+
  scale_fill_manual(values = c("#FFB300", "#FF6F00", "#43A047",
            "#2D62A3",  "#573794", "#28AAE1", "#870F0F", "#24b177",
            "#DDDDDD", "#7DCEA0", "#85C1E9", "#EB984E", "#EAD9D5", "#03CD4A",
            "#CDDC39", "#E0594B", "#C76CDE", "#8D6E63", "#486FF7", "#6300B5",
            "#88E200", "#012824", "#0D3290", "#A347FB", "#54FC7A", "#EB1388",
            "#B0978D", "#FE52CF", "#83F1F6", "#F1F847", "#2B1DFC", "#6C6F15", 
            "#6CA05C", "#7788CD", "#F502F3", "#0DC290", "#FA0E03", "#3CAA0A",
            "#BEFC8D", "#08F8EB", "#B1CD3F", "#D6A5FA", "#CE606C", "#AB1EBA",
            "#6ECC9F", "#054DDC", "#854F49", "#F22B21", "#3A0E43", "#225805",
            "#37D160", "#E4B974", "#A8BADE", "#47EDD1", "#F47A92", "#9106EB",
            "#81AA20", "#D7FDFD", "#5DEB2E", "#F82745", "#6435E0", "#027FFE",
            "#8E3101", "#16F648", "#1C15BC", "#8BE46E", "#8D6FA0", "#E68FC6",
            "#058CA9", "#9E018A", "#BDFD0B", "#B22760", "#2BF49F", "#CB9348",
            "#9D8303", "#C251A1", "#46ADAF", "#A3E3AF", "#22BB34", "#6EA3FA",
            "#260374", "#1C3854", "#405D37", "#C21DF3", "#FCEA92", "#537F88",
            "#FD4C18", "#F2D71E"))


#Saving plot

ggsave('fig_3.png', dpi = 350)
```

```{r}
##FIGURE 4 MASTER CODE

#Panel A

#Importing supplementary table S1

metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Identifying *B. pseudomycoides*-unique GCFs

bacillus = metadata %>% 
  group_by(gcf_id) %>% 
  mutate(
    Pseudomycoides_unique=ifelse(all(grepl('Bacillus_A_pseudomycoides', GTDB_Species)), "Yes","No")
  )

pseudo=bacillus %>% 
  filter(grepl('Yes', Pseudomycoides_unique))

#Importing the tree file, rooting and dropping outlier

tree=read.tree("pseudo_no_outliers_iqtree.treefile")
tree$tip.label<-if_else(str_detect(tree$tip.label, "^[0-9]"), paste0("X", tree$tip.label), tree$tip.label)
tree<-root(tree,"BTDB_2022-0000619.2")
tree=drop.tip(tree, "BTDB_2022-0000619.2")

#Setting up dataframe for tree annotation

tips=tree$tip.label

df=data.frame(tips)

tab=table(pseudo$BTyperDB_ID, pseudo$gcf_id)

df_2=as.data.frame.matrix(tab)

df_3=df_2 %>% 
  mutate_if(is.numeric, ~1 * (. > 0))

df$BTyper_ID <- sub("\\..*$", "", df$tips)

df_3$BTyper_ID=rownames(df_3)

df_4=merge(df, df_3, by="BTyper_ID", all.x=TRUE)


df_4[is.na(df_4)] <- 0

df_5=df_4[,-1]

rownames(df_5)=df_5$tips

df_6=df_5[,-1]

df_6[sapply(df_6, is.numeric)] <- lapply(df_6[sapply(df_6, is.numeric)], 
                                         as.factor)

#Making the tree plot

p=ggtree(tree,color="grey70")+
  geom_tippoint(size=0.5)+
  geom_treescale(x=0.0015, y=150)


p8=gheatmap(p, df_6, offset = 0.005, width = 1.5,
         colnames_position = "top", colnames_angle = 90, font.size = 2.5,
         colnames_offset_y = 7)+
  vexpand(.08, 1)+
  scale_fill_manual(values=c("grey95", "darkblue"), labels=c("Absent", "Present"), name="")

p8

#Saving the plot

ggsave('fig_4.png', dpi=350, width = 7.5, height = 10)


#Panel B

#This plot was made using LoVis4u, and later added in in Inkscape.
```

```{r}
##SUPPLEMENTARY FIGURE S1 MASTER CODE

#Setting up the dataframes

categories1=c("Ribosomal","NRP", "Terpene", "PKS", "Saccharide","NRP-PKS",
              "Other")
bgcs1=c(39912, 32729, 9699, 152, 2, 0,22353)
df1=data.frame(categories1, bgcs1)

categories2=c("Ribosomal","NRP","Terpene", "PKS", "Saccharide", "NRP-PKS",
              "Other")
bgcs2=c(9023, 25497, 43, 663,3564,2545,
        54014)
df2=data.frame(categories2,bgcs2)

#Making the bar plots

bar_anti=ggplot(data=df1, aes(x=categories1, y=bgcs1))+
  geom_bar(stat="identity", width = 0.5,
           color="#2A4765", fill="#2A4765")+ylab("# of BGCs")+xlab("")+theme(
             panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(),
             axis.text.x = element_text(size=11.5, angle=90),
             axis.line = element_line(colour = "black"),
             axis.title = element_text(size=11.5))

bar_gecco=ggplot(data=df2, aes(x=categories2, y=bgcs2))+
  geom_bar(stat="identity", width= 0.5,
           color="#2A4765", fill="#2A4765")+ylab("# of BGCs")+xlab("")+ theme(
             panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(),
             axis.text.x = element_text(size=11.5, angle=90),
             axis.line = element_line(colour = "black"),
             axis.title = element_text(size=11.5))

#Making and saving the figure

ggarrange(bar_anti, bar_gecco, align = "hv", ncol=2, nrow=1,
          labels = c("A", "B"))

ggsave('supp_fig_1.png', dpi=350, height = 5, width = 7)
```


```{r}
##SUPPLEMENTARY FIGURE S2 MASTER CODE

#Panel A

#Importing supplementary table S1
metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Making histogram showing genomes per GCF

genomes_per_gcf=metadata %>% 
  filter(!grepl('^BGC', cluster_id))  %>% 
  select(gcf_id, BTyperDB_ID) %>% 
  distinct()


genomes_per_gcf=genomes_per_gcf %>% 
  group_by(gcf_id) %>% 
  summarise(n=n())



p_hist_genomes=ggplot(genomes_per_gcf, aes(x=n))+
  geom_histogram(binwidth = 1, fill="#8acbcbff", colour="#2a4765ff")+
  theme_classic()+
  xlab("Genomes per GCF")+
  ylab("# of GCFs")

#Panel B

#Importing supplementary table S1

metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

#Making histogram showing panC per GCF

panC_per_gcf=metadata %>% 
  filter(!grepl('^BGC', cluster_id))  %>% 
  select(gcf_id, BTyper3_Adjusted_panC_Group.predicted_species.) 

panC_per_gcf$BTyper3_Adjusted_panC_Group.predicted_species. <- ifelse(grepl("\\*$", panC_per_gcf$BTyper3_Adjusted_panC_Group.predicted_species.), "Unknown", panC_per_gcf$BTyper3_Adjusted_panC_Group.predicted_species.)

panC_per_gcf$BTyper3_Adjusted_panC_Group.predicted_species.=sapply(panC_per_gcf$BTyper3_Adjusted_panC_Group.predicted_species., function(x)
  sub("Unknown(missing allele)", "Unknown",x,fixed=TRUE))


panC_per_gcf=panC_per_gcf %>% 
  distinct() %>% 
  group_by(gcf_id) %>% 
  summarise(n=n())

p_hist=ggplot(panC_per_gcf, aes(x=n))+
  geom_histogram(binwidth = 1, fill="#8acbcbff", colour="#2a4765ff")+
  theme_classic()+
  xlab("panC per GCF")+
  ylab("# of GCFs")


#Making the panel figure


ggarrange(
  p_hist_genomes,p_hist,
  align = "hv", ncol = 2, labels = c("A", "B")
)
ggsave('supp_fig_2.png', dpi = 350, height = 7, width = 12)

```

```{r}
##SUPPLEMENTARY FIGURE S3 MASTER CODE

antiSMASH   <- 503   
GECCO   <- 822  
Mix <- 390 

venn.plot <- draw.pairwise.venn(
  area1 = antiSMASH,
  area2 = GECCO,
  cross.area = Mix,
  category = c("antiSMASH", "GECCO"),
  fill = c("#AA3377", "#228833"),
  alpha = 0.5,
  cex = 2,
  cat.cex = 1.5,
  cat.pos = c(-20, 20)
)

grid::grid.newpage()
grid::grid.draw(venn.plot)

```

```{r}
##SUPPLEMENTARY FIGURE S5 MASTER CODE

#Import SeuratObject

pfams <- readRDS("btyperbgc_v1_0.55res_w_metadata.RDS")

#Make the clustree plot

clustree(pfams@meta.data, prefix = "peaks_snn_res.") 

#Export the figure

ggsave('supp_fig_5.png', dpi=350, height = 15, width = 8)


```

```{r}
##SUPPLEMENTARY FIGURE S6 MASTER CODE

#Import supplementary table S1

metadata=read.delim('supplementary_table_S1.tsv', header=TRUE)

bacillus = metadata %>% 
  group_by(gcf_id) %>% 
  mutate(
    Pseudomycoides_unique=ifelse(all(grepl('Bacillus_A_pseudomycoides', GTDB_Species)), "Yes","No")
  )

pseudo=bacillus %>% 
  filter(grepl('Yes', Pseudomycoides_unique))

#Import tree file, root and drop outgroup

tree=read.tree("pseudo_only_high_iqtree.treefile")
tree$tip.label<-if_else(str_detect(tree$tip.label, "^[0-9]"), paste0("X", tree$tip.label), tree$tip.label)
tree<-root(tree,"BTDB_2022-0000619.2")
tree=drop.tip(tree, "BTDB_2022-0000619.2")

#Set-up dataframe for annotation

tips=tree$tip.label

df=data.frame(tips)

tab=table(pseudo$BTyperDB_ID, pseudo$gcf_id)

df_2=as.data.frame.matrix(tab)

df_3=df_2 %>% 
  mutate_if(is.numeric, ~1 * (. > 0))

df$BTyper_ID <- sub("\\..*$", "", df$tips)

df_3$BTyper_ID=rownames(df_3)

df_4=merge(df, df_3, by="BTyper_ID", all.x=TRUE)


df_4[is.na(df_4)] <- 0

df_5=df_4[,-1]

rownames(df_5)=df_5$tips

df_6=df_5[,-1]

df_6[sapply(df_6, is.numeric)] <- lapply(df_6[sapply(df_6, is.numeric)], 
                                       as.factor)
#Make the plot

p=ggtree(tree,color="grey70")+
  geom_tippoint(size=0.5)+
  geom_treescale(x=0.0015, y=150)


p_tree_outliers_included=gheatmap(p, df_6, offset = 0.005, width = 1.5,
         colnames_position = "top", colnames_angle = 90, font.size = 2.5,
         colnames_offset_y = 7)+
  vexpand(.08, 1)+
  scale_fill_manual(values=c("grey95", "darkblue"), labels=c("Absent", "Present"), name="")

p_tree_outliers_included

ggsave('supp_fig_6.png',  dpi = 350, width = 9, height = 10)
```

